### PR TITLE
mds: fix SnapRealm::resolve_snapname for long name

### PR DESCRIPTION
**Pull request by ukernel**
Fixes: https://tracker.ceph.com/issues/38832
Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/mds/SnapRealm.cc [hunk 1]
```diff
@@ -454,15 +454,13 @@ snapid_t SnapRealm::resolve_snapname(std::string_view n, inodeno_t atino, snapid
...
-  //if (n[0] == '~') num = atoll(n.c_str()+1);
-
   bool actual = (atino == inode->ino());
   string pname;
   inodeno_t pino;
   if (n.length() && n[0] == '_') {
-    int next_ = n.find('_', 1);
-    if (next_ > 1) {
+    size_t next_ = n.find_last_of('_');
+    if (next_ != std::string::npos &&
```
Comment (batrick) [inline src/mds/SnapRealm.cc]: this check looks unnecessary because we already checked `n[0] == '_'`?

Comment (batrick) [inline src/mds/SnapRealm.cc]: this check looks unnecessary because we already checked `n[0] == '_'`?

File: src/mds/SnapRealm.cc [hunk 2]
```diff
@@ -454,15 +454,13 @@ snapid_t SnapRealm::resolve_snapname(std::string_view n, inodeno_t atino, snapid
...
+	next_ > 1 && next_ + 1 < n.length()) {
```
Comment (batrick) [inline src/mds/SnapRealm.cc]: Why not just do `n.back() == '_'`?

Comment (ukernel) [reply src/mds/SnapRealm.cc]: the format is _<snapname>_<inode number>

**Pull request merged**